### PR TITLE
HMAC Plugin

### PR DESCRIPTION
**Pull request by shashiranjan84**
HMAC Plugin

### CODE REVIEW

File: kong/plugins/hmac-auth/access.lua [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  end
+    
+  if conf.hide_credentials then
+    request.clear_header(header_name)
+  end
+  
+  return username, signature, algorithm
+end
+
+local function validate_signature(request, secret, signature, algorithm, defaultClockSkew)
```
Comment (harlow) [inline kong/plugins/hmac-auth/access.lua]: doesn't look like we're using the `algorithm` variable in this function. should it be used in the `-- validate signature` section or can it be removed as an argument?

Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: @harlow  for now only sha1 would be supported and would be mentioned in the manual, here argument is optional field and any value passed by user will be discarded.

Comment (harlow) [inline kong/plugins/hmac-auth/access.lua]: Makes sense :+1:

File: kong/plugins/hmac-auth/access.lua [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      end
+      return result
+    end)
+  end
+  return secret
+end
+
+function _M.execute(conf)
+  -- If both headers are missing, return 401
+  if not (ngx.req.get_headers()[AUTHORIZATION] or ngx.req.get_headers()[PROXY_AUTHORIZATION]) then
```
Comment (thibaultCha) [inline kong/plugins/hmac-auth/access.lua]: You are calling twice `ngx.req.get_headers()` (and once in `validate_signature` and another time in `retrieve_hmac_fields`). This is an expansive operation and the result should be cached.

Also you have `local ngx_set_header = ngx.req.set_header` but not the equivalent for `get_headers()`.

File: spec/plugins/hmac-auth/access_spec.lua [hunk 3]
```diff
@@ -0,0 +1,150 @@
+local spec_helper = require "spec.spec_helpers"
+local http_client = require "kong.tools.http_client"
+local cjson = require "cjson"
+local sha1 = require "sha1"
+local base64 = require "base64"
+
+local json_decode = cjson.decode
+local os_date = os.date
+local base64_encode = base64.encode
```
Comment (thibaultCha) [inline spec/plugins/hmac-auth/access_spec.lua]: It is not useful to perform global optimisations on tests, it only makes them less readable.

File: kong/plugins/hmac-auth/access.lua (hunk 2 above)
Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: sorry missed it, updating it.

File: spec/plugins/hmac-auth/access_spec.lua (hunk 3 above)
Comment (shashiranjan84) [inline spec/plugins/hmac-auth/access_spec.lua]: Just to understand how its fine to use global optimisations at one but non readable at another place. isn't it better to be optimise it everywhere given alias used for function is very descriptive?

Comment (thibaultCha) [inline spec/plugins/hmac-auth/access_spec.lua]: Global optimisation is only useful when the code is intensively run. Tests run only once in a (long) while.

Comment (shashiranjan84) [inline spec/plugins/hmac-auth/access_spec.lua]: For sake of consistency why not use it everywhere, anyway it's not going impact the real time application performance?
Anyway i ll update the code, i made a major logic change to include other header as part of signature. Sending a PR soon

**Pull request merged**